core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -103,46 +103,41 @@
   /** All objects in the set. This set is required to guarantee uniqueness of objects. */
   // TODO(gpang): remove this set, and just use the indexes.
   private final ConcurrentHashSet<T> mObjects = new ConcurrentHashSet<>(8, 0.95f, 8);
-  /**
-   * Map from {@link FieldIndex} to the index. An index is a map from index value to set of
-   * objects with that index value.
-   */
-  private final Map<FieldIndex<T>, ConcurrentHashMap<Object, ConcurrentHashSet<T>>> mIndexMap;
 
   /**
-   * An interface representing an index for this {@link IndexedSet}, each index for this set must
-   * implement the interface to define how to get the value of the field chosen as the index. Users
-   * must use the same instance of the implementation of this interface as the parameter in all
-   * methods of {@link IndexedSet} to represent the same index.
-   *
-   * @param <T> type of objects in this {@link IndexedSet}
+   * Map from index definition to the index. An index is a map from index value to one or a set of
+   * objects with that index value. A unique index is an index where each index value only maps to
+   * one object. A non-unique index is an index where an index value can map to one or more objects.
    */
-  public interface FieldIndex<T> {
-    /**
-     * Gets the value of the field that serves as index.
-     *
-     * @param o the instance to get the field value from
-     * @return the field value, which is just an Object
-     */
-    Object getFieldValue(T o);
-  }
+  private final Map<IndexDefinition<T>, FieldIndex<T>> mIndices;
 
   /**
    * Constructs a new {@link IndexedSet} instance with at least one field as the index.
    *
-   * @param field at least one field is needed to index the set of objects
-   * @param otherFields other fields to index the set
+   * @param firstIndexDefinition at least one field is needed to index the set of objects
+   * @param otherIndexDefinitions other index definitions to index the set
    */
   @SafeVarargs
-  public IndexedSet(FieldIndex<T> field, FieldIndex<T>... otherFields) {
-    Map<FieldIndex<T>, ConcurrentHashMap<Object, ConcurrentHashSet<T>>> indexMap =
-        new HashMap<>(otherFields.length + 1);
-    indexMap.put(field, new ConcurrentHashMap<Object, ConcurrentHashSet<T>>(8, 0.95f, 8));
-    for (FieldIndex<T> fieldIndex : otherFields) {
-      indexMap.put(fieldIndex, new ConcurrentHashMap<Object, ConcurrentHashSet<T>>(8, 0.95f, 8));
+  public IndexedSet(IndexDefinition<T> firstIndexDefinition, IndexDefinition<T>...
+      otherIndexDefinitions) {
```

Comment:
@aaudiber, that's a good suggestion! But Gene and I are planning to remove the mObjects of the IndexedSet class and use the first index of IndexedSet to perform the role of mObjects. Under that circumstance, to change the signature to `IndexedSet(IndexDefinition<T>... indexDefinitions)` seems cannot simplify the code because we should do extra check in a few functions then.